### PR TITLE
[bug-fix] Set number of threads based on allocated CPU count in Docker containers

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

PyTorch by default sets number of threads to 1/2 available cores. However, in Docker containers, the built-in Python methods return the number of cores of the host machine, not the container. This PR reads the allocated CPU count from `cgroup` and uses that to set the num threads. 

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by chriselion** [comment]
Nice catch. I've run into this before and didn't think about it for pytorch.

File: ml-agents/mlagents/torch_utils/cpu_utils.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    if period > 0 and quota > 0:
+        return int(quota // period)
+    else:
+        return os.cpu_count()
+
+
+def _read_in_integer_file(filename: str) -> int:
+    try:
+        with open(filename) as f:
+            return int(f.readlines()[0])
```
Comment (chriselion) [inline ml-agents/mlagents/torch_utils/cpu_utils.py:32]: There's some code on https://bugs.python.org/issue36054 - they use the equivalent of `int( f.read().rstrip() )`

**Review by vincentpierre — APPROVED** [review]

File: ml-agents/mlagents/torch_utils/cpu_utils.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/torch_utils/cpu_utils.py:32]: I like the `read().rstrip()` better that what I had, changed

**Review by chriselion — APPROVED** [review]

File: ml-agents/mlagents/torch_utils/cpu_utils.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    """
+    Gets the number of threads to use. For most problems, 4 is all you
+    need, but for smaller machines, we'd like to scale to less than that.
+    By default, PyTorch uses 1/2 of the available cores.
+    """
+    num_cpus = _get_num_cpus()
+    return max(min(num_cpus // 2, 4), 1) if num_cpus is not None else None
+
+
+def _get_num_cpus() -> Optional[int]:
```
Comment (chriselion) [inline ml-agents/mlagents/torch_utils/cpu_utils.py:16]: nit: maybe `_get_num_available_cpus()`?

**Review by chriselion — APPROVED** [review]

**Pull request merged**